ReText/singleapplication.py

```diff
@@ -0,0 +1,167 @@
+from PyQt5.QtCore import QSystemSemaphore
+from PyQt5.QtNetwork import QLocalServer
+from PyQt5.QtNetwork import QLocalSocket
+
+class SingleApplication(QObject):
+	"""
+	SingleApplication is a class that wrap around the single application
+	framework.
+	"""
+
+	# Modes
+	(
+		# Server mode indicated that we started the first application
+		Server,
+		# Client mode means that another App already started, we should just
+		# exit or send some message to that App.
+		Client,
+	) = range(0, 2)
+
+	# Signals
+
+	receivedMessage = pyqtSignal(bytes)
+
+	def __init__(self, name, parent=None):
+		QObject.__init__(self, parent)
+		self._name = name
+		self._mode = self.Server
+		self._sharedMemory = QSharedMemory(self._name)
+		self._systemSemaphore = QSystemSemaphore(self._name + "-2", 1)
+		self._server = None
+		self._client = None
+		self._localSockets = {}
+
+		# Fix for *nix:
+		# http://stackoverflow.com/questions/5006547/qt-best-practice-for-a-single-instance-app-protection
+		# http://habrahabr.ru/post/173281/
+
+		self._systemSemaphore.acquire();
+		try:
+			unixFix = QSharedMemory(self._name)
+			unixFix.attach()
+			unixFix.detach()
+			unixFix = None
+		finally:
+			self._systemSemaphore.release();
+
+	@property
+	def name(self):
+		return self._name
+
+	@property
+	def mode(self):
+		return self._mode
+
+	def _onLocalSocketReadyRead(self, localSocket):
+		if localSocket.bytesAvailable() <= 0:
+			return
+
+		self._localSockets[localSocket] += localSocket.readAll()
+		data = self._localSockets[localSocket]
+		if len(data) > 4:
+			# First 4bytes is an native Integer.
+			dataSize = struct.unpack("@I", data[:4])[0]
+			receivedDataSize = len(data) - 4
+			if receivedDataSize < dataSize:
+				return
+
+			self.receivedMessage.emit(bytes(data[4:]))
+
+			# Remove the command socket
+			del self._localSockets[localSocket]
+			localSocket.deleteLater()
+
+	def _onServerNewConnection(self):
+		while self._server.hasPendingConnections():
+			localSocket = self._server.nextPendingConnection()
+			self._localSockets[localSocket] = b""
+			localSocket.readyRead.connect(functools.partial(
+				self._onLocalSocketReadyRead, localSocket))
+
+	def start(self):
+		# Ensure we run only one application
+		isAnotherRunning = False
+		if not self._sharedMemory.isAttached():
+			# If share memory is attachable, that means there have another
+			# running application
+			self._systemSemaphore.acquire()
+			try:
+				isAnotherRunning = self._sharedMemory.attach()
+				if isAnotherRunning:
+					self._sharedMemory.detach()
+			finally:
+				self._systemSemaphore.release()
+
+		if not isAnotherRunning:
+			# If we already attached by previous attach test, we will create
+			# the share memory.
+			self._systemSemaphore.acquire()
+			try:
+				isAnotherRunning = not self._sharedMemory.create(1, QSharedMemory.ReadWrite)
+			finally:
+				self._systemSemaphore.release()
+
+			# If there something happened during create procedure lead us can't
```

Comment:
Maybe we don't need to comment here to tell what's going on, because we must do the clear job if the function failed to create, right?

I means, no matter what it happening, maybe system error or else, it must be clear after failed, so, we clear it.